[4.0] Select image/audio/video/document file from Media Manager

dgrammatiko opened the pull request:
Pull Request for Issue https://github.com/joomla/joomla-cms/discussions/32642 and https://www.facebook.com/groups/joomlanospam/posts/10157777796480997/?__cft__[0]=AZXC8G44TJLPO0QeEbytZwSVaZX2YpNFHjOUcx1xUqyJjBUZGqgh763LWT-Lx2xKyvJfIm-LQO6wUquCrqneOLqSAlx3AOpRDAxnMrliJHTTK1AIkT4Mh-BVfJCJjMVFEyDYelefuWE8UMA3X6_-XUFRasmSPT8VNhUVlOln0Ujj4LCrwnBYH_QLfColkn2QEkk&__tn__=%2CO%2CP-R

Apparently, the Joomla community finds it totally fine to use software that is **based on a javascript that's 2 major versions behind**. It's not fine for me thus the PR...

### Summary of Changes
- Reworked the media select Javascript
- Changed the one field for all the supported extensions to multiple fileds in Media Manager (this needs to be communicated as it's a B/C break, and also it needs updates to the sql files)
- Reworked/added file types in the vue App (the files need to enable the preview)

Please extract this 
[Archive.zip](https://github.com/joomla/joomla-cms/files/6720757/Archive.zip)
 to the images directory, or use your own mp3/mp4/pdf files

### Testing Instructions
- Apply the PR or downlad the installable from github

- If you're applying the PR then go to the Media manager options and 
- Make sure the legal mime types are: `image/jpeg,image/gif,image/png,image/bmp,audio/mp3,video/mp4,application/msword,application/excel,application/pdf,application/powerpoint,text/plain,application/x-zip`
- the video extensions are `mp4`
- the audio extensions are `mp3`
- the document extensions are `odg,odp,ods,odt,pdf,png,ppt,txt,xcf,xls,csv`
- and finally the images extensions are: `bmp,gif,jpg,jpeg,png,ico` **This is IMPORTANT**
- 
- Go to tinyMCE and:

- Select the image from the CMS Content dropdown
- Then select the video file, click on the controls and embed checkboxes, change the width and height and then click select image at the bottom of the modal
- Press couple returns type `Video Link` and select the text you just wrote
- Then select the video file, without selecting the controls and embed checkboxes (width and height is irrelevant here) and then click select image at the bottom of the modal
- Check the produced HTML by clicking the toggle button under the editor

- Select the image from the CMS Content dropdown
- Then select the audio file, click on the controls and embed checkboxes and then click select image at the bottom of the modal
- Press couple returns type `Audio Link` and select the text you just wrote
- Then select the audio file, without selecting the controls and embed checkboxes and then click select image at the bottom of the modal
- Check the produced HTML by clicking the toggle button under the editor

- Select the image from the CMS Content dropdown
- Then select the PDF file, click on the embed checkboxe  change the width and height and then click select image at the bottom of the modal
- Press couple returns type `PDF Link` and select the text you just wrote
- Then select the PDF file, without selecting the embed checkboxe and then click select image at the bottom of the modal
- Check the produced HTML by clicking the toggle button under the editor

- your final html should look like:
```html
<p><video controls="controls" width="800" height="600">
<source src="/images/flower.mp4" type="video/mp4" /></video></p>
<p> </p>
<p><a href="images/flower.mp4" download="">Video Link</a></p>
<p> </p>
<p><audio src="images/t-rex-roar.mp3" controls="controls"></audio></p>
<p> </p>
<p><a href="images/t-rex-roar.mp3" download="">Audio Link</a></p>
<p> </p>
<p><object data="images/core_xy_exploded_infographic1.pdf" type="application/pdf" width="800" height="600">
You don't have a pdf plugin, but you can <a download="" href="/images/core_xy_exploded_infographic1.pdf"> download the pdf file.</a></object></p>
<p> </p>
<p><a href="images/core_xy_exploded_infographic1.pdf" download="">PDF Link</a></p>
```

- Check that the insertion of an image is still ok
- Check that the media FIELDS (intro/fulltext images) are still functional
- Finally save the article and visit it in the front end. ~That's it.~ 

- Edit the definition of the intro image https://github.com/joomla/joomla-cms/blob/7f5dd8ee5f6a690a88f05926bef6cb8456055bb8/administrator/components/com_content/forms/article.xml#L735-L739 by adding `types="images,audios"`
- Test the field, you should be able now to select images and audio files, also try to upload an unsupported file (eg video and pdf)
- Repeat with different combinations `types="images,audios,videos,documents"`


### Actual result BEFORE applying this Pull Request

Impossible to select/embed anything other than an image file


### Expected result AFTER applying this Pull Request

It's ok to select/embed audio, video, document files
<img width="865" alt="Screenshot 2021-06-26 at 23 04 08" src="https://user-images.githubusercontent.com/3889375/123526089-da132d00-d6d5-11eb-9451-3465805d6b89.png">
<img width="1133" alt="Screenshot 2021-06-26 at 23 04 38" src="https://user-images.githubusercontent.com/3889375/123526090-db445a00-d6d5-11eb-98ce-eb875a5857c7.png">
<img width="721" alt="Screenshot 2021-06-26 at 23 04 28" src="https://user-images.githubusercontent.com/3889375/123526091-dbdcf080-d6d5-11eb-86d3-4fcf685a427e.png">
<img width="857" alt="Screenshot 2021-06-26 at 23 04 18" src="https://user-images.githubusercontent.com/3889375/123526094-dbdcf080-d6d5-11eb-9eea-79cf447976d4.png">

Front end:
![j4 local_index php_teststststtsts](https://user-images.githubusercontent.com/3889375/123526488-89510380-d6d8-11eb-80fe-a2dbe85d6b4a.png)

### Documentation Changes Required

There are few todos that I skipped here:
- sql files
- remove renamed files
- rename the xtd button image to media as the image is confusing now

@wilsonge @laoneo any feedback?

dgrammatiko committed:
meh

brianteeman commented:
looking forward to testing this tomorrow

dgrammatiko committed:
Remove test files

dgrammatiko committed:
Fix inconsistencies

dgrammatiko committed:
install sqls

brianteeman commented on administrator/components/com_media/src/Model/ApiModel.php:
```diff
@@ -533,17 +533,52 @@ private function isMediaFile($path)
+						',',
+						ComponentHelper::getParams('com_media')->get(
+						'image_extensions',
+						'bmp,csv,doc,gif,ico,jpg,jpeg'
```
doc isnt an image ;)

denysdesign replied:
Somebody drawing or embedding image in doc 😏😂

dgrammatiko replied:
It's called text art 😂. Obviously copy/paste mistake

brianteeman commented:
Tested using your supplied files. Only the pdf uploads. The mp3 and mp4 give a generic cant be uploaded message.
Tested a larger 5mb mp4 and it doesnt appear to do anything. No upload, no error, nothing
Sometimes after one of the failures the session is killed and you have to log in again

brianteeman commented:
The problem was in your test instructions. Better to clear the existing values in the options and then save in order to get the defaults. If it had been customised at all it would fail

brianteeman commented:
Click on a pdf file and then click on the three dots
Check the console
```
media-manager.min.js?6a2a37958931d4cd9934c0d049a970a8:1 Uncaught (in promise) TypeError: Cannot read property 'focus' of undefined
    at Proxy.<anonymous> (media-manager.min.js?6a2a37958931d4cd9934c0d049a970a8:1)
```

brianteeman commented:
mp4 files have previews but there is no preview icon (you need to double click on the file to see the preview)

pdf files do not have previews but you can still double click on the file and the previous item is displayed

brianteeman commented:
If you dont check the box "Use native elements" before inserting then nothing added to the article

richard67 commented:
For files and folder deletion on update with script.php you can leave it to @wilsonge and me. One of us will run the tool to generate the lists of files and folders when this PR has been merged and before the next release after that. So it does not need to be done in this PR. But if you want to do it here because you wanna make it right, I can help later.

For SQL scripts we have to decide how much we want to touch the media manager options on update. Since it will be at least risky and hacky trying to add single values to the json, we might have to completely replace the params so people would lose their settings on update if they have changed some. I can help with making the update SQL but not with 
making the decision if we want that or not.

brianteeman commented:
@richard67 whichever is decided on the sql there should be a postinstallation message. either saying the settings have been reset to the defaults or saying that you can update the settings to add support for pdf, etc

richard67 commented:
> @richard67 whichever is decided on the sql there should be a postinstallation message. either saying the settings have been reset to the defaults or saying that you can update the settings to add support for pdf, etc

Yeah, that makes sense. Personally I'd prefer the "We don't touch the settings and let the postinstall message tell people that they can update their settings to add support for PDF and so on" approach.

It can be that we have already an update SQL touching these parameters when updating from 3.10. If this is the case, we could modify that one to set them right. I have to check if there is one.

brianteeman commented:
IIRC we discussed this when I added webp support (or tried to)

dgrammatiko committed:
update

richard67 commented:
I've just checked: It seems there is no update SQL script where we already touch the params of the com_media extension. So there is no need to modify such a script.

dgrammatiko commented:
> If you dont check the box "Use native elements" before inserting then nothing added to the article

Fixed, there are 3 modes: embed, insert asa link with predefined text, insert as a link with selected text from the editor (the 2nd was missing)

> mp4 files have previews but there is no preview icon (you need to double click on the file to see the preview)

The preview needs refactoring so it can handle all the file types, also I thought that removing the magnifying glass I removed the preview but I totally forgot the double click


> Click on a pdf file and then click on the three dots
> Check the console

Some side effect from copy/pasting, investigating

> The problem was in your test instructions. Better to clear the existing values in the options and then save in order to get the defaults. If it had been customised at all it would fail

Will update the instructions for existing installations

A comment about SQL files and B/C:
- Since J4 supports multiple local adapters the `images` adapter should be sniffed on `update` time and the default directory for media files should be `media-uploads` or something similar. Having audio, video or other files under a directory `images` is weird but maybe I'm heretic here. My point is that this can be done in a non B/C breaking way with a post install script for updates. Also doing it this way means that we get rid of runtime code that we have right now to do the adjustment, sounds more solid solution to me anyways.
Anyways I'm ok whatever the decision is here, post install message or not or whatever...

richard67 commented:
> A comment about SQL files and B/C:
> 
>     * Since J4 supports multiple local adapters the `images` adapter should be sniffed on `update` time and the default directory for media files should be `media-uploads` or something similar. Having audio, video or other files under a directory `images` is weird but maybe I'm heretic here. My point is that this can be done in a non B/C breaking way with a post install script for updates. Also doing it this way means that we get rid of runtime code that we have right now to do the adjustment, sounds more solid solution to me anyways.
>       Anyways I'm ok whatever the decision is here, post install message or not or whatever...

@dgrammatiko It would be good to have that post install script or procedure. But that would maybe be better a new routine in the script.php than an update SQL script. If necessary @joomdonation could have a look on it.

dgrammatiko committed:
Reverse the strings

dgrammatiko committed:
Fix console err on ellipsis click

dgrammatiko committed:
oops

dgrammatiko committed:
Fix previews

dgrammatiko committed:
meh

dgrammatiko committed:
meh

brianteeman commented:
Preview of video now working correctly

Preview of pdf - almost - the preview opens but its black. Clicking on rotate makes it appear (maybe that will give a clue)

Audio file will not upload

brianteeman committed:
Update plg_editors-xtd_image.ini

brianteeman committed:
Update plg_editors-xtd_image.sys.ini

dgrammatiko commented:
> Audio file will not upload

Make sure the legal mime types are: `image/jpeg,image/gif,image/png,image/bmp,audio/mp3,video/mp4,application/msword,application/excel,application/pdf,application/powerpoint,text/plain,application/x-zip`

brianteeman commented:
Checked that twice before posting

`image/jpeg,image/gif,image/png,image/bmp,audio/mp3,video/mp4,application/msword,application/excel,application/pdf,application/powerpoint,text/plain,application/x-zip`

brianteeman commented:
Article intro and full image
This works when selecting an image (as before)
However you can select a non image file - instead of some type of error message saying "this isnt an image etc" it just silently returns

dgrammatiko commented:
>  it just silently returns

Yes, I couldn't decide what would be the best scenario for this case so I silently ignored anything non-image. This might need an alert or something similar.

> Checked that twice before posting

There's something fishy as the tests are also failing for no obvious reason

richard67 commented:
@dgrammatiko Here the screenshot from drone from system tests: https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/34634/system-tests/45401/MediaListCest.showsDefaultFilesAndFolders.mysql.fail.png

It says "syntax error, unexpected ')'".

Here the part of the log with the test where it fails:

```
MediaListCest: Test that it shows the joomla default media files and folders.
89s
325Signature: MediaListCest:showsDefaultFilesAndFolders
89s
326Test: tests/Codeception/acceptance/administrator/components/com_media/MediaListCest.php:showsDefaultFilesAndFolders
89s
327Scenario --
89s
328 I do administrator login 
89s
329  [Cookies] [{"name":"0cd014c79b960a69d3b33ac5245b0833","value":"2c944nm5pml9uatbrlgsjngk43","path":"/","domain":"localhost","expiry":1624893164,"secure":false,"httpOnly":false}]
89s
330  [Snapshot] Restored "ci-admin" session snapshot
89s
331 I create directory "images/test-dir"
89s
332   I get suite configuration 
89s
333   I get suite configuration 
89s
334 Created /tests/www/test-install/images/test-dir
89s
335 I am on page "administrator/index.php?option=com_media&..."
89s
336  [GET] http://localhost/test-install/administrator/index.php?option=com_media&path=local-images:/
89s
337 I wait for media loaded 
90s
338   I wait for element {"class":"media-loader"},3
90s
339   I wait for element not visible {"class":"media-loader"}
90s
340   I wait 1
90s
341 I see contents ["banners","headers","sampledata","joomla_black.png","powered_by.png"]
91s
342   I see element {"class":"media-browser-item"}
91s
343 I delete directory "images/test-dir"
91s
344   I get suite configuration 
91s
345 Deleted /tests/www/test-install/images/test-dir
91s
346 I execute js "window.sessionStorage.removeItem("joomla...."
91s
347  Screenshot and page source were saved into '/********/src/tests/Codeception/_output/' dir
91s
348 FAIL 
```

Does that help?

dgrammatiko committed:
meh

dgrammatiko committed:
erm

dgrammatiko commented:
> Does that help?

Yup, turns out PHP is far away from becoming a functional programming language...

dgrammatiko committed:
undo this

dgrammatiko committed:
Implement a show-on

richard67 commented:
@dgrammatiko Now only analysis4x is failing in drone, which is some security stuff using RIPS and is not related to your PR.

There was a general problem with that, but that should be solved now. So if you wanna get rid of the red test and see all green, just update your branch to latest 4.0-dev. But it's not absolutely necessary because we know the drone failure here is that thing.

dgrammatiko committed:
nope

dgrammatiko committed:
For the media field display only images

dgrammatiko commented:
> Article intro and full image
> This works when selecting an image (as before)
> However you can select a non image file - instead of some type of error message saying "this isnt an image etc" it just silently returns

Now the media field will display **only** images (this is configurable in the code and can be adjusted to fit any combo)

> Audio file will not upload

For this, I need some help, maybe @laoneo ?

Fedik commented:
Looks good to me (except broken upload for audio/video).

I would suggest to add `webm, mov` for video, and `ogg` for audio, from beginning. Or later we will get a complains about it.

dgrammatiko commented:
> Looks good to me (except broken upload for audio/video).

Technically this bug is unrelated to this PR, it just happens to become obvious from the changes here 

> I would suggest to add webm, mov for video, and ogg for audio, from beginning. Or later we will get a complains about it.

Yeah, also `webp` images and maybe `avif` (avif is already supported on GD and Imagick but right now you have to compile from source to enable support, but will be available in some next minor release)

brianteeman commented:
There is already an open PR for webp. Its just stuck in an open state because cropper.js doesnt work in firefox for windows.

dgrammatiko commented:
@brianteeman can you do a simple check: in your apache configuration there should be a file `mime.types`, could you compare it with this one 
[mime.types.txt](https://github.com/joomla/joomla-cms/files/6725305/mime.types.txt) (backup your existing, remove the `.txt` and place it in the same folder).
I have no clue if php is sniffing the mime types from apache but if it does then there isn't a problem but rather a misconfiguration (I can upload mp3/mp4 without any problems here)

brianteeman commented:
Rebuilding my site at the moment but I think I may have found the problem

mime.types sas 
`audio/mpeg					mpga mp2 mp2a mp3 m2a m3a`

This PR says
`audio\mp3`

Checking the [spec](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types) `audio\mpeg`

So the problem does appear to be related to the pr and not the server

dgrammatiko commented:
Good catch!

dgrammatiko committed:
Fix/add mime types, extensions

dgrammatiko committed:
Add also webm

brianteeman commented:
I can now add mp3 

What is the intended behaviour with the mp3 preview. I just get a filename
![image](https://user-images.githubusercontent.com/1296369/123623113-0083b600-d805-11eb-8eda-7852ab756568.png)

brianteeman commented:
![image](https://user-images.githubusercontent.com/1296369/123623146-07aac400-d805-11eb-8f1c-e8f251cfc8f4.png)

I find this confusing. I expected the description to be an explanation of the option. The last sentence really describes what to do if you dont want a d/l link.

I would probably have two options
**Embed**
_Use native elements audio, video or object_

**Download**
_Use a download link_

dgrammatiko commented:
> What is the intended behaviour with the mp3 preview. I just get a filename

Weird, you should get: 
<img width="1783" alt="Screenshot 2021-06-28 at 12 41 14" src="https://user-images.githubusercontent.com/3889375/123623936-4abd6500-d80e-11eb-8040-90ba5f274ff1.png">

and a markup: 

<img width="630" alt="Screenshot 2021-06-28 at 12 41 34" src="https://user-images.githubusercontent.com/3889375/123623982-57da5400-d80e-11eb-8b6d-91f8959e3e85.png">

Does your apache `mime.types` has uncommented the mp3 part? (I think you can drop a mimes.type in the root of the joomla folder, eg grab the one from a couple of comments above)

dgrammatiko commented:
> I would probably have two options

You mean something like a switcher or use one of the 2 proposals?

brianteeman commented:
Weird - the preview works now

> You mean something like a switcher or use one of the 2 proposals?

switcher/checkbox/radio - dont care but something that means the user is actively selecting embed or download

N6REJ commented:
Nice to finally see this.

dgrammatiko committed:
Checkbox to radio

dgrammatiko commented:
> switcher/checkbox/radio - dont care but something that means the user is actively selecting embed or download

I went for a radio and it looks like:
![Screenshot 2021-06-28 at 14 35 17](https://user-images.githubusercontent.com/3889375/123637494-5fa1f480-d81e-11eb-93dd-32864fd0595e.png)
![Screenshot 2021-06-28 at 14 35 26](https://user-images.githubusercontent.com/3889375/123637499-60d32180-d81e-11eb-96db-ff6c1055769b.png)

dgrammatiko committed:
remove console

dgrammatiko committed:
pfff CS

dgrammatiko committed:
Sync the default values

dgrammatiko committed:
Reposition the details element

brianteeman commented:
Is there a reason you went for a fixed size embed instead of a responsive embed? (https://embedresponsively.com/)

dgrammatiko commented:
> Is there a reason you went for a fixed size embed instead of a responsive embed?

The wide/height are not for sizing but rather (as with images) to inform the browser what's the aspect ratio (plain English to allow the browser to correctly calculate the empty space till the video/object is loaded)

For example the code that linked does this (creates an empty 16/9 parent div and then inserts the iframe into it. This is a hack, we don't need to do this anymore, browsers respect and retrieve the correct aspect ratio from the width/height attributes. 

A good resource for width/height from Jen Simmons: https://speakerdeck.com/jensimmons/using-the-html-width-and-height-attributes-to-improve-web-page-loading

brianteeman commented:
I have tested this item :white_check_mark: successfully on 08ce8176e10e9f52f49b97d9491cd36f78677277<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34634">issues.joomla.org/tracker/joomla-cms/34634</a>.</sub>

Quy commented on administrator/components/com_media/config.xml:
```diff
@@ -56,6 +48,16 @@
+			size="50"
+			default="bmp,gif,jpg,jpeg,png,ico,mp3,mp4,odg,odp,ods,odt,pdf,png,ppt,txt,xcf,xls,csv"
+			showon="restrict_uploads:1"
+			/>
```
```suggestion
		/>
```

Quy replied:
```suggestion
		/>
```

dgrammatiko committed:
Update administrator/components/com_media/config.xml

Co-authored-by: Quy <quy@fluxbb.org>

Quy commented on administrator/components/com_media/tmpl/file/default.php:
```diff
@@ -25,41 +25,47 @@
+	'csrfToken'          => Session::getFormToken(),
+	'uploadPath'         => $this->file->path,
+	'editViewUrl'        => Uri::base() . 'index.php?option=com_media&view=file' . ($tmpl ? '&tmpl=' . $tmpl : '') . $imgOnly,
+	'imagesExtensions'   => explode(',', $params->get('image_extensions', 'bmp,gif,jpg,jpeg,png,ico')),
```
```suggestion
	'imagesExtensions'   => explode(',', $params->get('image_extensions', 'bmp,gif,jpg,jpeg,png,webp')),
```

dgrammatiko committed:
Change the order of the radios

dgrammatiko commented:
> Maybe swap Embed and Download radio buttons since Enable Controls relates to Embed.

Done

Quy commented on administrator/components/com_media/tmpl/file/default.php:
```diff
@@ -25,41 +25,47 @@
+	'imagesExtensions'   => explode(',', $params->get('image_extensions', 'bmp,gif,jpg,jpeg,png,ico')),
+	'audioExtensions'    => explode(',', $params->get('audio_extensions', 'mp3')),
+	'videoExtensions'    => explode(',', $params->get('video_extensions', 'mp4')),
+	'documentExtensions' => explode(',', $params->get('document_extensions', 'odg,odp,ods,odt,pdf,png,ppt,txt,xcf,xls,csv')),
```
```suggestion
	'documentExtensions' => explode(',', $params->get('doc_extensions', 'odg,odp,ods,odt,pdf,png,ppt,txt,xcf,xls,csv')),
```

Quy commented on administrator/components/com_media/tmpl/file/default.php:
```diff
@@ -25,41 +25,47 @@
+	'csrfToken'          => Session::getFormToken(),
+	'uploadPath'         => $this->file->path,
+	'editViewUrl'        => Uri::base() . 'index.php?option=com_media&view=file' . ($tmpl ? '&tmpl=' . $tmpl : '') . $imgOnly,
+	'imagesExtensions'   => explode(',', $params->get('image_extensions', 'bmp,gif,jpg,jpeg,png,ico')),
+	'audioExtensions'    => explode(',', $params->get('audio_extensions', 'mp3')),
+	'videoExtensions'    => explode(',', $params->get('video_extensions', 'mp4')),
```
Sync default values?

dgrammatiko replied:
Done

dgrammatiko committed:
Remove listeners on disconected

dgrammatiko committed:
Sync more default values

Quy commented:
Edit an article.
Click CMS Content > Media
Click on an MP3 file.
Click `Embed` radio button.
Click `Enable Controls` checkbox.
The checkbox should not disappear.

dgrammatiko committed:
oops

dgrammatiko commented:
> The checkbox should not disappear.

Fixed, thanks

Quy commented:
Checked/unchecked `Enable Controls` produce the same markup.

`<p><audio src="images/t-rex-roar.mp3" controls="controls"></audio></p>`

dgrammatiko committed:
fix controls checkbox

dgrammatiko commented:
> Checked/unchecked Enable Controls produce the same markup.

Fixed, thanks

Quy commented:
Missing icon in list view.

![34634-icon](https://user-images.githubusercontent.com/368084/123682776-4be79580-d800-11eb-8a2e-21736e0d4498.jpg)

alikon commented:
can you take a look at #34314 @dgrammatiko  please

dgrammatiko committed:
icon + typo

dgrammatiko commented:
> Missing icon in list view.

Fixed

> can you take a look at #34314

I've seen it already but I don't think I'm the right guy for feedback there. My feeling is that we might want to control the contents of a folder that is returned (mind there's no permissions as in other components so things might have to be done different, eg utilising the same technic as: https://github.com/joomla/joomla-cms/pull/34634/files#diff-8404917e9476fe232ef9c49920ced95e9ce5825e5efd60f955ecde240a2a090aL533-R583 ) or wait till a db layer is added in com_media (or I'm totally wrong...).

Quy commented:
Select an MP3 file.
Click `Embed` radio button.
Don't click `Enable Controls`.

Here is the markup, but nothing is displayed.

`<audio src="images/t-rex-roar.mp3"></audio>`

dgrammatiko committed:
Remove the controls opt, tags always get controls

dgrammatiko commented:
> Here is the markup, but nothing is displayed.
> `<audio src="images/t-rex-roar.mp3"></audio>`

Option to add or not controls removed. To my defend here, if you have a JS player you only need the tag with the source as the rest are often controlled by the JS player. Anyhow, if in the future people asking for this option the code is here...

Quy commented on administrator/language/en-GB/com_media.ini:
```diff
@@ -48,6 +52,8 @@ COM_MEDIA_FIELD_PATH_FILE_FOLDER_LABEL="Path to Files Folder"
 COM_MEDIA_FIELD_PATH_IMAGE_FOLDER_LABEL="Path to Images Folder"
 COM_MEDIA_FIELD_RESTRICT_UPLOADS_DESC="Restrict uploads for lower than manager users to images if Fileinfo or MIME Magic isn't installed."
+COM_MEDIA_FIELD_RESTRICT_UPLOADS_EXTENSIONS_DESC="Restrict uploads for lower than manager users to files in the list."
+COM_MEDIA_FIELD_RESTRICT_UPLOADS_EXTENSIONS_LABEL="Allowed extensions"
```
```suggestion
COM_MEDIA_FIELD_RESTRICT_UPLOADS_EXTENSIONS_LABEL="Allowed Extensions"
```

dgrammatiko committed:
Update administrator/language/en-GB/com_media.ini

Co-authored-by: Quy <quy@fluxbb.org>

brianteeman commented:
Embedding an object requires and alt  or equivalent see 
https://accessibilityinsights.io/info-examples/web/object-alt/

brianteeman commented:
The preview of a video is different to a preview of an image with regards to width and the position of the close button. I am assuming its some css

![image](https://user-images.githubusercontent.com/1296369/123748417-15ffeb00-d8ac-11eb-9b2e-633c7e49491b.png)

![image](https://user-images.githubusercontent.com/1296369/123748422-19937200-d8ac-11eb-83b8-90e15ad3b516.png)

dgrammatiko committed:
Allow media field to use any media type

This just introduces an attribute `types`
```
	/**
	 * Comma separated types of files for Media Manager
	 * Possible values: images,audios,videos,documents
	 *
	 * @var    string
	 * @since  __VERSION_DEPLOY__
	 */
	protected $types;
```

- It needs changes in the markup and JS to support preview for different types

A PR will follow once (if) this is merged as the testing here will get too complicated…

dgrammatiko commented:
> Embedding an object requires and alt or equivalent see

I will implement this later on today

dgrammatiko committed:
CS

dgrammatiko committed:
Add title to object tag

dgrammatiko committed:
CS

dgrammatiko commented:
> The preview of a video is different to a preview of an image with regards to width and the position of the close button. I am assuming its some css

The current implementation is using inline style background for the image, thus the difference

dgrammatiko committed:
Really now?

dgrammatiko committed:
oops

dgrammatiko committed:
few more fixes

dgrammatiko committed:
stop failing

dgrammatiko committed:
center the preview modal

dgrammatiko committed:
Better defaults

dgrammatiko committed:
Nice try…

Quy commented:
Close icon overlaps filename.

![34634-close](https://user-images.githubusercontent.com/368084/123875384-6dbf4600-d8ee-11eb-856a-d65ae7af67ae.jpg)

micker commented:
Hello i tested work very good for me, just 2 notices :
- the additionnal windows doesn't close when we close media
- it will be cool to open media popup when we do right click or double click on media easier to change option
thanks a million @dgrammatiko Now we have a real MEDIA manager !!!

dgrammatiko committed:
Select audio,video,document

dgrammatiko committed:
update some dependencies

dgrammatiko committed:
sync +

dgrammatiko committed:
Support all media types on Media Field

A new attribute is introduced named `types`
The value can be comma separated with possible entries:
- images
- audios
- videos
- documents

Default (no types provides) falls back to images only
Side effect you can’t upload audio,video,document files from a media field with type images…

dgrammatiko committed:
The cs is hiddeous

dgrammatiko committed:
grrr

dgrammatiko committed:
lol

dgrammatiko commented:
> the additionnal windows doesn't close when we close media

I don't understand what you mean, the media manager opens in a modal, not a new window...

> it will be cool to open media popup when we do right click or double click on media easier to change option

This is not possible for 2 main reasons:
- there is always an extra step after selecting a file (provide alt text or embed or link for download) so it will never be one-step process
- there are 2 client-side applications: the media manager and a tiny Custom Element that asks for the extra info. These 2 cannot be combined into one as the one runs in an iframe and the other one on the main document. To get everything in the same window requires a huge refactor of few parts (doable but not a couple of days work...)

dgrammatiko committed:
Fix the close button on previews

micker commented:
Ok for point 2
For point sorry for my Bad explain i mean the small zone with option (embed or download) this zone doesnt close when WE close media

dgrammatiko commented:
>  i mean the small zone with option (embed or download) this zone doesnt close when WE close media

Can you record a video demonstrating what's not working? Seems fine here

dgrammatiko committed:
Clinking on the left tree should close the

Additional Data Custom Element

dgrammatiko committed:
Select all checkbox should also close

the Additional Data Custom Element

dgrammatiko committed:
Breadcrumb should also close

the Additional Data Custom Element

dgrammatiko committed:
CS

micker commented:
i do small video, that the additionnal option zone
https://www.loom.com/share/b378c2a3c7304547ad577516f65c4742
thnaks

dgrammatiko commented:
> i do small video, that the additionnal option zone

Ok, I see what you mean and already fixed 3 cases that the aditional data should be dismissed but as long as some file is selected this will be visible. You can click on the title `Aditional Data` though, and that will be minimised. I can't do anything more here

micker commented:
Ok that already better !

dgrammatiko commented:
@micker can you please mark you test on the issues site?

brianteeman commented:
I have tested this item :white_check_mark: successfully on 42d28ef15eb33ab497af98ef671762a4039dcb56<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34634">issues.joomla.org/tracker/joomla-cms/34634</a>.</sub>

micker commented:
I have tested this item ✅ successfully on 42d28ef too

richard67 commented:
@micker Could you please go to the issue tracker here and mark your test result in the right way? Just copying the message from another test is not sufficient. G to https://issues.joomla.org/tracker/joomla-cms/34634 , use the "Test this" button at the top left corner, select the appropriate test result and then submit. Thanks in advance.

micker commented:
I have tested this item :white_check_mark: successfully on 42d28ef15eb33ab497af98ef671762a4039dcb56<br /><br />Tested and perfect
Sorry Richard a mistake<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34634">issues.joomla.org/tracker/joomla-cms/34634</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34634">issues.joomla.org/tracker/joomla-cms/34634</a>.</sub>

wilsonge commented:
This seems to important to not merge right now given it has b/c implications. Although it's not great given our point of RC. Thankyou so much @dgrammatiko for your hard work on this!

brianteeman commented:
woohoo!!!

dgrammatiko commented:
Thanks

@wilsonge the B/C break is the change of some of the com_media params and to be honest here even what I did here is not perfect. An ideal solution is to have a data structure like
```json
{
  supported: [
    {
      extension: 'png',
      mine: 'image/png',
      preview: true,
      upload: true
    }
  ]
}
```
That will allow to remove the hardcoded preview extensions in the com_media app (vue) and also if done right (subforms?) it will be easier in the backend (eg have a pool that users enable/disable, instead of typing tings)

Kostelano commented on api/language/en-GB/joomla.ini:
```diff
@@ -241,11 +241,22 @@ JFIELD_LOGOUT_REDIRECT_URL_LABEL="Logout Redirect"
+JFIELD_MEDIA_WIDTH_LABEL="Width"
+JFIELD_MEDIA_TITLE_LABEL="Title"
+; Do not translate the text between the {}
+JFIELD_MEDIA_UNSUPPORTED="You don't have a {extension} plugin, but you can {tag} download the {extension} file.</a>"
```
I see that it is combined. But is everything true in this line? Confused by the end of the line.

@dgrammatiko

richard67 replied:
Hmm, yes ... where does that closing anchor tag belong to?

Kostelano replied:
There is also the same ending in another line. Copy / paste

richard67 replied:
@Kostelano Maybe you can make an issue, or if you have more time make a PR to fix it? It would need to get that text shown and then check the original markup. Browser inspection tools might repair the invalid markup and so not show any mistake, so it could be tricky.

dgrammatiko replied:
The first part is here https://github.com/joomla/joomla-cms/blob/83e3dd33a2853a72939fb4e3d682f90ee2ad414d/build/media_source/system/js/fields/joomla-media-select.w-c.es6.js#L235

richard67 replied:
I see ... not nice, but ok ... thanks for reporting back.

dgrammatiko replied:
I just realized that the closing tag shouldn't be in the ini file. Well, my bad 😞

richard67 replied:
You can make a PR ... but it should be  soon because there might be language freeze for J4 in near future.

Quy commented on administrator/components/com_media/resources/scripts/components/browser/items/document.vue:
```diff
@@ -0,0 +1,221 @@
+    <div class="media-browser-item-preview">
+      <div class="file-background">
+        <div class="file-icon">
+          <span class="fas fa-file-pdf" />
```
Can you please improve this to reflect their respective icons instead of a PDF icon for all document types? Thanks.

dgrammatiko replied:
What would be the appropriate icon? Also you can PR the change yourself nothing too fancy here 😉

Quy replied:
pdf icon for .pdf
word icon for .doc
excel icon for .xls
otherwise default to generic file icon for the other extensions

richard67 replied:
Don’t forget docx, docm, xlsx and xlsm

brianteeman replied:
You can do this in pure css so that the markup doesnt have to contain everything. See this example https://gschoppe.com/uncategorized/add-auto-sensing-file-type-icons-to-lists-of-downloads-with-fontawesome-and-css/

dgrammatiko replied:
So change line 10 to ` <span v-bind:class="getFileClass" />`
and add after line 44 (after the data(){} ):

```js
  computed: {
    /* Get the hashed URL */
    getFileClass: (item) => {
      switch (item.mime_type) {
        case 'application/pdf':
          return 'fas fa-file-pdf';
        case 'application/msword':
        case 'application/vnd.openxmlformats-officedocument.wordprocessingml.document':
          return 'fas fa-file-word';
        case 'application/vnd.ms-excel':
        case 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet':
          return 'fas fa-file-excel';
        case 'application/vnd.ms-powerpoint':
        case 'application/vnd.openxmlformats-officedocument.presentationml.presentation':
          return 'fas fa-file-powerpoint';
        default:
          return 'fas fa-file';
      }
    },
  },
```

Modify it to your taste

dgrammatiko replied:
> You can do this in pure css so that the markup doesnt have to contain everything.

Actually, on the web, the extensions are totally FOOBAR. The only way to tell what a type of a file really is, is by checking the mime type. The mime type is not available to the HTML/CSS unless the application actually passes it either as a class or a data attribute. 

What I'm saying is that you can have a totally valid PDF (or whatever) without any extension. Actually, Joomla should never ever do anything based on the extensions as this is a huge SECURITY risk. 

/rant

Quy replied:
> and add after line 44 (after the data(){} ):

Should it be after line 177 in the same file?

dgrammatiko replied:
yes: https://github.com/joomla/joomla-cms/blob/b84fc74c882b0b86d220bf6c5b9999e9c1a17958/administrator/components/com_media/resources/scripts/components/browser/items/document.vue#L177
